Multiple changes in the appearance of ultidisplay

green-striped-gecko opened the pull request:
- progress bar now in the message line during print (working on an on/off switch stored in the eeprom)
- auto home with 5 mm up at the end (needs to be fixed for not inversed printers)
- reversed sorting of files on the SD-card (makes it quicker to print the file that was added last)
- timer stops at the end of print (most people would like to know how long it took to print)
- added a visual "<" sign at the end of submenus where a value can be selected, so users know where they are

Hi Daid,
I deleted the fork and remade it.
Hopefully this time there are only the files that are needed in the pull request.
All not necessary files are not committed from my local repro (README.md and configuration.h) , so I think it should be fine.
I also reversed the finished printing back to Partial release. I need to find the right place for it.

Cheers, greengecko

ErikZalm committed:
Fixed E_ENABLE_PIN errors

green-striped-gecko committed:
Multiple changes in the visualisation of ulticontroller

- progress bar now in the message line during print (working on an on/off switch stored in the eprom for progressbar)
- auto home with 5 mm up at the end (needs to be fixed for not inversed printers)
- reversed sorting of files on the SD-card (makes it quicker to print the file that was added last)
- timer stops at the end of print (most people would like to know how long it took to print)
- added a visual "<" sign at the end of submenus where a value can be selected, so users know where they are

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -395,7 +410,7 @@ void MainMenu::showStatus()
     uint16_t time=millis()/60000-starttime/60000;
     
-    if(starttime!=oldtime)
+    if(starttime!=oldtime && card.sdprinting==true) //stop timer at the end of print!!
```
Add this to make sure time is stopping after print, as this is the time you want to know after the print.

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -453,6 +473,16 @@ void MainMenu::showStatus()
     lcd.print(itostr3((int)percent));
     lcdprintPGM("%SD");
+  
+ if (percent>0) 
```
Progress bar added to the message text.

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -528,7 +561,7 @@ void MainMenu::showPrepare()
       break;
     case ItemP_home:
-      MENUITEM(  lcdprintPGM(MSG_AUTO_HOME)  ,  BLOCK;enquecommand("G28");beepshort(); ) ;
+      MENUITEM(  lcdprintPGM(MSG_AUTO_HOME)  ,  BLOCK;enquecommand("G28");enquecommand("G1 X0 Y0 Z5 E0");beepshort(); ) ;
```
Go up 5 mm after homing. Not sure how to check for different printer which need -5mm instead?

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -602,13 +635,16 @@ void MainMenu::showAxisMove()
                     if(linechanging)
                     {
 			enquecommand("G91");
+                        lcd.setCursor(19,line);lcdprintPGM("<");
```
Insert a visual clue once a submenu that allows to change a value is entered. This is followed by the according delete command (via empty space).

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -2306,8 +2393,8 @@ void MainMenu::showSD()
         {
-          card.getfilename(i-FIRSTITEM);
-          //Serial.print("Filenr:");Serial.println(i-2);
+          card.getfilename(nrfiles + 1-i);
```
Revers the order of files

green-striped-gecko commented on Marlin/ultralcd.pde:
```diff
@@ -2323,7 +2410,7 @@ void MainMenu::showSD()
         {
           BLOCK
-          card.getfilename(i-FIRSTITEM);
+          card.getfilename(nrfiles +1-i);
```
And once selected, print the according file. The order could be also implemented as a setup in a submenu (in addition to progressbar) to custimize the appearance

daid commented:
Looking a lot better :-) You did replace the readme file, so that was also included in the pull request. No biggy. 

I think I can try out the changes tomorrow and merge them then.

I think the "card.getfilename(nrfiles +1-i);" in ultralcd.pde should be "card.getfilename((nrfiles - 1) - (i - FIRSTITEM));"
As (i - FIRSTITEM) is the actual index, and you want to substract that from (nrfiles - 1) because at index 0 you want the last item, which is (nrfiles - 1).
It results in the same final result, but it reads a bit easier.

green-striped-gecko commented:
sounds fine to me and I can see the point to make it more readable. Just to be picky, would that not mean more RAM is taken (one byte I guess or is that compiled into a single constant? One other question: Would it help to simplify the ulticontroller menu part by creating subfunctions and get rid of repeated chunks of code to save RAM. Again I admit I have not a really good understanding what goes into memory and what into the flash part of the arduino.

daid commented:
Programming rule nr1: Do not optimize. You are not smarter then the compiler optimizer.

The compiler will optimize the statement, so there wouldn't be extra constants in the code, and there won't be extra instructions executed. Even if it did, it would use extra flash, not RAM. As the code is put in flash, not in ram. (Important to remember for these chips, as the RAM is very small)

I have ideas to cleanup the LCD code. And actually also all the other Marlin code as well. But the LCD code really needs an overhaul, as it's hard to read and change. And contains lots of duplicate code.

green-striped-gecko committed:
Changed reverse sorting following daids suggestion for better readability

(nrfiles +1 -i)    to (nrfiles-1) - (i-FIRSTITEM)

green-striped-gecko committed:
config before daids last version

green-striped-gecko committed:
Merge branch 'origin/Marlin_v1' of https://github.com/green-striped-gecko/Marlin.git into Marlin_v1

Conflicts:
	Marlin/Configuration.h

Signed-off-by: green-striped-gecko <bernd@yellowgecko.de>

green-striped-gecko committed:
Change color submenu added

now you can stop the print at a predefined layer, the nozzle goes to 190 190
an M0 is called (code basically used from daids thingiverse example) and restarted
at the previous position.

green-striped-gecko commented:
HI daid,

this change adds a change color submenu to the main menu. It allows the user to specify a layer height and if he wants to change the filament. If so code such as your thingiverse sample is issued (go to 190/190 and M0 to wait for a button push).

green-striped-gecko committed:
Updated Color Change

now changes at exactly the layer height >= instead of >.
Also returns to Main_Status (and does not timeout) important if someone wants
to stop exactly when he returns to Main_Status pressing Watch (exit).

green-striped-gecko commented:
Hi daid 
I know you are probably busy though i am curious if you think my pull requests will be at least partly updatec into your marlin build. If you do not like the coding thats okay ( i will not be offended)  but it would great if you could tell me if you plan to have a look at it and potentially get a feedback or not.

Cheers, Bernd

daid commented:
There is nothing wrong with your code, I just haven't found the time to work on Firmware yet. I'm very busy with Cura 12.10, after that I want to look at the firmware.

Small note on your latest change, the color change. I think it's a good addition, but I would call it "auto-pause at height". Because you could also use it to insert objects into your print next to changing colors.

(Your code base is also a bit behind my version, which is why I can no longer auto-merge)

green-striped-gecko committed:
now also configuration.h is changed...

green-striped-gecko committed:
Merge branch 'Marlin_v1' of https://github.com/daid/Marlin.git into Marlin_v1

Conflicts:
	Marlin/Configuration.h
	Marlin/ultralcd.pde

green-striped-gecko committed:
change color change to autopause

suits better as suggested by daid as the feature can be used
to change colors and also for other issues.

green-striped-gecko commented:
Hy Daid, 
I followed your suggestion (I did not think about other features as e.g. putting something into print) and renamed everything to autopause.
Not sure if it helps or makes things worse now -  I fetched and merged your latest version into my version before pushing this one. I had to resolve only a very minor conflict (some character in ultracd.pde were coded line by line and not as hexcode).

daid commented:
I can merge again now, so that's good. I like the rename :) I'll review it better later on.

green-striped-gecko commented:
Just saw that sailfish has the same feature calling it auto pause @ zpos i think. Or a bit exaggerating multicolor with a single nozzle. As if there was only one color before....

green-striped-gecko commented:
Hi Daid,

Cura 12.10 is Great work. I wondered what version of Marlin comes with it?

Cheers, Bernd

daid commented:
Same as all the previous versions, because it's proven to be very stable and we didn't want to risk a new version.

green-striped-gecko commented:
Thanks good to know.

Cheers, Bernd

# 

Dr Bernd Gruber       Tel: (02) 6206 3804 Fax: (02) 6201 2328
Assistant Professor
Institute for Applied Ecology
Faculty of Applied Science
University of Canberra   ACT 2601 AUSTRALIA
Email: bernd.gruber@canberra.edu.aumailto:bernd.gruber@canberra.edu.au
WWW: http://www.canberra.edu.au/faculties/science/staff/profiles/dr-bernd-gruber

Australian Government Higher Education Provider Number CRICOS:#00212K

NOTICE & DISCLAIMER: This email and any files transmitted with it may contain
confidential or copyright material and are for the attention of the addressee
only. If you have received this email in error please notify us by email
reply and delete it from your system. The University of Canberra accepts
no liability for any damage caused by any virus transmitted by this email.

# 

On 11/11/2012, at 9:51, "daid" <notifications@github.com<mailto:notifications@github.com>> wrote:

Same as all the previous versions, because it's proven to be very stable and we didn't want to risk a new version.

—
Reply to this email directly or view it on GitHubhttps://github.com/daid/Marlin/pull/6#issuecomment-10260943.
